File: internal/source/damodaran/ratings.go
package damodaran

import (
	"context"
	"errors"
	"fmt"
	"github.com/yinhm/alphalake/internal/source/reference"
	"math/big"
)

const CreditDataset = "synthetic-credit-large-2026-v1"
const CreditURL = "https://pages.stern.nyu.edu/~adamodar/New_Home_Page/datafile/ratings.html"
const CreditScript = "internal/source/damodaran/ratings.py"
const CreditParserVersion = "damodaran-ratings-large-2026-v1"

type CreditBand struct {
	Lower         string `json:"lower"`
	Upper         string `json:"upper"`
	Rating        string `json:"rating"`
	RawValue      string `json:"raw_value"`
	Value         string `json:"value"`
	SourceLocator string `json:"source_locator"`
}
type CreditSnapshot struct {
	reference.Header
	Observations []CreditBand `json:"observations"`
}

func ParseCredit(ctx context.Context, python, script, path string) (CreditSnapshot, string, error) {
	var s CreditSnapshot
	hash, err := reference.Parse(ctx, python, script, path, nil, &s)
	if err != nil {
		return s, "", err
	}
	return s, hash, ValidateCredit(s)
}
func ValidateCredit(s CreditSnapshot) error {
	if err := s.Header.Validate("alphalake-credit-spreads-v1", CreditParserVersion); err != nil {
		return err
	}
	ratings := []string{"D2/D", "C2/C", "Ca2/CC", "Caa/CCC", "B3/B-", "B2/B", "B1/B+", "Ba2/BB", "Ba1/BB+", "Baa2/BBB", "A3/A-", "A2/A", "A1/A+", "Aa2/AA", "Aaa/AAA"}
	lower := []string{"-100000", "0.2", "0.65", "0.8", "1.25", "1.5", "1.75", "2", "2.25", "2.5", "3", "4.25", "5.5", "6.5", "8.50"}
	upper := []string{"0.199999", "0.649999", "0.799999", "1.249999", "1.499999", "1.749999", "1.999999", "2.2499999", "2.49999", "2.999999", "4.249999", "5.499999", "6.499999", "8.499999", "100000"}
	if s.ObservationDate != "2026-01-01" || len(s.Observations) != 15 {
		return errors.New("unsupported credit scope/month")
	}
	for i, o := range s.Observations {
		if o.Rating != ratings[i] || o.Lower != lower[i] || o.Upper != upper[i] || o.SourceLocator != fmt.Sprintf("table[1]/tr[%d]/td[1:4]", i+5) {
			return errors.New("unsupported credit band")
		}
		if err := reference.Decimal12(o.RawValue, o.Value, 100); err != nil {
			return err
		}
		v, ok := new(big.Rat).SetString(o.Value)
		if !ok || v.Sign() < 0 || v.Cmp(big.NewRat(1, 1)) >= 0 {
			return errors.New("invalid credit spread")
		}
	}
	return nil
}

File: internal/source/damodaran/ratings.py
"""仅解析 2026-01 大型非金融企业合成评级表；源区间和百分数逐字保留。"""
from html.parser import HTMLParser
from decimal import Decimal
from pathlib import Path
import hashlib
import json
import platform
import sys


class Tables(HTMLParser):
    def __init__(self):
        super().__init__(); self.tables=[]; self.rows=None; self.row=None; self.cell=None; self.text=[]
    def handle_starttag(self, tag, attrs):
        if tag=='table':
            if self.rows is not None: raise ValueError('nested table')
            self.rows=[]
        elif tag=='tr' and self.rows is not None: self.row=[]
        elif tag in ('td','th') and self.row is not None: self.cell=[]
    def handle_data(self, data):
        self.text.append(data)
        if self.cell is not None: self.cell.append(data)
    def handle_endtag(self, tag):
        if tag in ('td','th') and self.cell is not None:
            self.row.append(' '.join(''.join(self.cell).split())); self.cell=None
        elif tag=='tr' and self.row is not None:
            self.rows.append(self.row); self.row=None
        elif tag=='table' and self.rows is not None:
            self.tables.append(self.rows); self.rows=None


def parse(raw):
    p=Tables(); p.feed(raw.decode('mac_roman')); p.close()
    if 'Data used is as of January 2026' not in ' '.join(' '.join(p.text).split()):
        raise ValueError('unreviewed analysis month')
    targets=[(i,t) for i,t in enumerate(p.tables,1) if any('For large non-financial service firms' in ' '.join(r) for r in t)]
    if len(targets)!=1: raise ValueError('missing/duplicate nonfinancial table')
    table,rows=targets[0]
    if not any(r[:4]==['>','≤ to','Rating is','Spread is'] for r in rows):
        raise ValueError('unsupported boundary headers')
    ratings=['D2/D','C2/C','Ca2/CC','Caa/CCC','B3/B-','B2/B','B1/B+','Ba2/BB','Ba1/BB+','Baa2/BBB','A3/A-','A2/A','A1/A+','Aa2/AA','Aaa/AAA']
    observations=[]
    for i,row in enumerate(rows,1):
        if len(row)<4 or row[2] not in ratings: continue
        if len(row)!=9 or not row[3].endswith('%'): raise ValueError('unsupported band row')
        lo,hi=map(Decimal,row[:2]); spread=Decimal(row[3][:-1])
        if not lo.is_finite() or not hi.is_finite() or not spread.is_finite() or lo>=hi or not 0<=spread<100:
            raise ValueError('invalid band')
        observations.append(dict(lower=row[0],upper=row[1],rating=row[2],raw_value=row[3][:-1],value=f'{spread/100:.12f}',source_locator=f'table[{table}]/tr[{i}]/td[1:4]'))
    if [r['rating'] for r in observations]!=ratings: raise ValueError('incomplete rating scope')
    if any(Decimal(a['upper'])>=Decimal(b['lower']) for a,b in zip(observations,observations[1:])):
        raise ValueError('overlapping bands')
    return dict(contract='alphalake-credit-spreads-v1',observation_date='2026-01-01',sha256=hashlib.sha256(raw).hexdigest(),
                parser_version='damodaran-ratings-large-2026-v1',runtime='python='+platform.python_version(),observations=observations)


if __name__=='__main__':
    print(json.dumps(parse(Path(sys.argv[1]).read_bytes()),allow_nan=False))
